### DISCUSSION TITLE
Semihosting with `SVC instruction` or `Breakpoint instruction`?

### DISCUSSION THREAD

**Discussion by bat42** [picolibc/picolibc#225] [Q&A] [answered]
Hello, 

is semihosting (https://github.com/picolibc/picolibc/blob/main/doc/os.md#semihosting-support) implemented as `SVC instruction` or `Breakpoint instruction`. 

In chapter 21 of https://www.segger.com/downloads/jlink/UM08001, it's mentioned that both variants have there pro(s) and con(s) and I wanted to know which one is used.

Thanks for the help

BTW, semihosting is running well on my `-march=armv7e-m` MCU

**Update**

I read

```
Note
ARM processors use the SVC instructions, formerly known as SWI instructions, to make semihosting calls. 
However, if you are compiling for an ARMv6-M or ARMv7-M, for example a Cortex-M1 or Cortex-M3 processor, 
semihosting is implemented using the BKPT instruction.
```
form https://www.keil.com/support/man/docs/armcc/armcc_pge1358787046598.htm, which means on my platform `Breakpoint instructions` are used by default?

**Comment by keith-packard** [comment, marked as answer]
picolibc uses the recommended instructions on each architecture, so "it depends". On all M profile chips, that's `bkpt #0xab`. On all other thumb chips, it's `svc #0xab`. On all other chips, it's `svc #0x123456`. You can see all of this in [semihost-arm.S](https://github.com/picolibc/picolibc/blob/main/semihost/machine/arm/semihost-arm.S). Glad semihosting is working for you.
